source/val/validation_state.cpp

```diff
@@ -425,6 +426,25 @@ bool ValidationState_t::RegisterUniqueTypeDeclaration(
     key.insert(key.end(), inst.words + words_begin, inst.words + words_end);
   }
 
+  // If module has extension SPV_KHR_variable_pointers, then OpTypePointer
+  // is allowed to create identical types so long as their decorations differ.
```

Comment:
Looking again, this is not the rule.
The SPV_KHR_variable_pointers extension says 
> Pointer types are also allowed to have multiple id's for the same opcode and operands, to allow for differing ArrayStride Array Stride decoration values.

> Also, in that paragraph, "non-aggregate types" will then generally be "non-aggregate non-pointer types."

From 2.8 in the SPIR-V spec:
> It is valid to declare multiple aggregate type <id>s having the same opcode and operands. This is to allow multiple instances of aggregate types with the same structure to be decorated differently. (Different decorations are not required; two different aggregate type ids are allowed to have identical declarations and decorations, and will still be two different types.)

That means the validator should allow distinct OpTypePointer with the same operands, and with the same decorations.

Reply:
So I think all this new code boils down to:
```
if (inst.opcode == SpvOpTypePointer && HasExtension....) return true;
```
Even better:  put the check up in the callee, along with the aggregate types.  That is, modify TypeUniquePass to do the pointer-and-has-extension check.